sno/gpkg_adapter.py

```diff
@@ -0,0 +1,246 @@
+        "geometry_type_name": type_name,
+        "srs_id": gpkg_spatial_ref_sys[0]["srs_id"],
+        "z": z,
+        "m": m,
+    }
+
+
+def generate_gpkg_spatial_ref_sys(dataset):
+    """Generate a gpkg_spatial_ref_sys meta item from a dataset."""
+    geom_columns = _get_geometry_columns(dataset.schema)
+    if not geom_columns:
+        return []
+
+    srs_pathname = geom_columns[0].extra_type_info["geometrySRS"]
+    definition = dataset.get_srs_definition(srs_pathname)
+    return wkt_to_gpkg_spatial_ref_sys(definition)
+
+
+DEFAULT_GPKG_SPATIAL_REF_SYS = [
+    {
+        "srs_name": "Unknown CRS",
+        "definition": "",
+        "organization": "EPSG",
+        "srs_id": 0,
+        "organization_coordsys_id": 0,
+        "description": None,
+    }
+]
+
+
+def wkt_to_gpkg_spatial_ref_sys(wkt):
+    """Given a WKT srs definition, generate a gpkg_spatial_ref_sys meta item."""
+    return _gpkg_spatial_ref_sys(SpatialReference(wkt), wkt)
+
+
+def osgeo_to_gpkg_spatial_ref_sys(spatial_ref):
+    """Given an osgeo SpatialReference, generate a gpkg_spatial_ref_sys meta item."""
+
+    return _gpkg_spatial_ref_sys(spatial_ref, spatial_ref.ExportToWkt())
+
+
+def _gpkg_spatial_ref_sys(spatial_ref, wkt):
+    spatial_ref.AutoIdentifyEPSG()
+    return [
+        {
+            "srs_name": spatial_ref.GetName(),
+            "definition": wkt,
+            "organization": spatial_ref.GetAuthorityName(None),
+            "srs_id": spatial_ref.GetAuthorityCode(None),
+            "organization_coordsys_id": spatial_ref.GetAuthorityCode(None),
+            "description": None,
+        }
+    ]
+
+
+DEFAULT_SRS_STR = "EPSG:0"
+
+
+def wkt_to_srs_str(wkt):
+    """Given a WKT srs definition, generate a sensible name for it."""
+    return osgeo_to_srs_str(SpatialReference(wkt))
+
+
+def osgeo_to_srs_str(spatial_ref):
+    """Given a osgeo SpatialReference, generate a sensible name for it."""
+    auth_name = spatial_ref.GetAuthorityName(None)
+    auth_code = spatial_ref.GetAuthorityCode(None)
+    return f"{auth_name}:{auth_code}"
+
+
+def generate_sqlite_table_info(dataset):
+    """Generate a sqlite_table_info meta item from a dataset."""
+    return [_columnschema_to_gpkg(i, col) for i, col in enumerate(dataset.schema)]
+
+
+def _get_geometry_columns(schema):
+    return [c for c in schema.columns if c.data_type == "geometry"]
+
+
+def gpkg_to_v2_schema(
+    sqlite_table_info, gpkg_geometry_columns, gpkg_spatial_ref_sys, id_salt
+):
+    """Generate a v2 Schema from the given gpkg meta items."""
+    return Schema(
+        [
+            _gpkg_to_column_schema(
+                col, gpkg_geometry_columns, gpkg_spatial_ref_sys, id_salt
+            )
+            for col in sorted(sqlite_table_info, key=_sort_by_cid)
+        ]
+    )
+
+
+def _sort_by_cid(sqlite_col_info):
+    return sqlite_col_info["cid"]
+
+
+def _gpkg_to_column_schema(
+    sqlite_col_info, gpkg_geometry_columns, gpkg_spatial_ref_sys, id_salt
```

Comment:
THis one's complicated enough that I think it deserves a docstring

Reply:
Done